### ISSUE TITLE
Prediction input dimension error

### ISSUE THREAD

**Issue by junyang-cynopsis** [idealo/image-super-resolution#27]
Hi im trying to do inference on the rrdn model i just trained but i received the following error

Error when checking input: expected LR_input to have shape (40, 40, 3) but got array with shape (908, 908, 3)

May I check whats the reason to be (40,40,3) input dimension?

**Comment by cfrancesco** [comment]
When doing inference you should not specify an input size when creating a model instance. For example
`model = RRDN(arch_params={'C':4, 'D':3, 'G':32, 'G0':32, 'T':10, 'x':2})`

**Comment by junyang-cynopsis** [comment]
You're right, removed the patch size and model predicts successfully. Another issue arise now when i train the model, i receive similar error which comes at random phase of the training epochs

**ValueError: Error when checking input: expected input_3 to have shape (90, 90, 3) but got array with shape (90, 89, 3)**

**Comment by morph3us-net** [comment]
Hello, please do not mind me hijackingthis thread, but I have the same error message.
I simply just copied together the documentation like this (where I, after having read this thread, added the optional "patch_size" argument to explicitly set it to "None"). So I am clueless what I am doing wrong. Could somebody please help? :)

`
import numpy as np
from ISR.models import RDN
from PIL import Image

img = Image.open('input/test.png')
lr_img = np.array(img)

rdn = RDN(arch_params={'C':3, 'D':10, 'G':64, 'G0':64, 'x':2}, patch_size=None)
rdn.model.load_weights('weights/rdn-C3-D10-G64-G064-x2_PSNR_epoch134.hdf5')

sr_img = rdn.predict(lr_img)
upscaled_image = Image.fromarray(sr_img)
upscaled_image.save("output\\upscaled.png")`

**Comment by junyang-cynopsis** [comment]
> Hello, please do not mind me hijackingthis thread, but I have the same error message.
> I simply just copied together the documentation like this (where I, after having read this thread, added the optional "patch_size" argument to explicitly set it to "None"). So I am clueless what I am doing wrong. Could somebody please help? :)
> 
> `
> import numpy as np
> from ISR.models import RDN
> from PIL import Image
> 
> img = Image.open('input/test.png')
> lr_img = np.array(img)
> 
> rdn = RDN(arch_params={'C':3, 'D':10, 'G':64, 'G0':64, 'x':2}, patch_size=None)
> rdn.model.load_weights('weights/rdn-C3-D10-G64-G064-x2_PSNR_epoch134.hdf5')
> 
> sr_img = rdn.predict(lr_img)
> upscaled_image = Image.fromarray(sr_img)
> upscaled_image.save("output\upscaled.png")`

Try removing the patchsize input parameter all together for inference
ie **rrdn  = RRDN(arch_params={'C':4, 'D':3, 'G':64, 'G0':64, 'T':10, 'x':scale})**

Updates for training own model, i set a if condition when input batch is erroneous, it will skip the particular training step. Hopefully it will ignore dimensional errors. However, still not sure what is causing the issue

**Comment by morph3us-net** [comment]
Thank you very much for your reply! Unfortunately it doesnt help... :(

I added the explicit "patch_size" argument only after encountering the error.
Its the same, with and without it.

I also do not want to train my own model, just upscale a few images... (just to clear that up)

**Comment by junyang-cynopsis** [comment]
@morph3us-net Does lr_img = np.array(img) successfully converts your image into numpy array? Is your image path valid?

**Comment by morph3us-net** [comment]
Yes, it does, printing out lr_img shows a numpy array.
I have now tried the exact same thing with a much lower resolution jpg, instead of a bigger .png and now it mysteriously works...
Maybe it has something to do with how many channels there are in a file (e.g. RGB vs. RGBA)?

**Comment by cfrancesco** [comment]
> You're right, removed the patch size and model predicts successfully.
> Another issue arise now when i train the model, i receive similar error which comes at random phase of the training epochs
> 
> **ValueError: Error when checking input: expected input_3 to have shape (90, 90, 3) but got array with shape (90, 89, 3)**

Hi, sorry for the late reply. 
Do you have images that are smaller than 90 on the 2nd axis in your training set? 
Try with a smaller patch size or remove the images that are smaller than the patch size you are trying to use.
There should definitely be a check for this though..

**Comment by cfrancesco** [comment]
> Yes, it does, printing out lr_img shows a numpy array.
> I have now tried the exact same thing with a much lower resolution jpg, instead of a bigger .png and now it mysteriously works...
> Maybe it has something to do with how many channels there are in a file (e.g. RGB vs. RGBA)?
> 
> Edit:
> I added a
> `img = img.convert("RGB")`
> as second line and it seems to work now.
> 
> P.S.: Looking back at your error message, it seems that your images are already only RGB, so that doesn't help you, sorry...

Very strange, PIL should have a format that works with ISR. 
If you don't mind, could you post the exact original error that you encountered?

**Comment by wanfuse123** [comment]
Getting same error. Sorry to take over thread! Any solution yet?

Traceback (most recent call last):
  File "LoadandPredict.py", line 12, in <module>
    rdn = RDN(arch_params={'C':6, 'D':20, 'G':64, 'G0':64, 'x':2})
  File "/home/steven/.local/lib/python2.7/site-packages/ISR/models/rdn.py", line 66, in __init__
    self.model = self._build_rdn()
  File "/home/steven/.local/lib/python2.7/site-packages/ISR/models/rdn.py", line 165, in _build_rdn
    )(LR_input)
  File "/home/steven/.local/lib/python2.7/site-packages/keras/engine/base_layer.py", line 431, in __call__
    self.build(unpack_singleton(input_shapes))
  File "/home/steven/.local/lib/python2.7/site-packages/keras/layers/convolutional.py", line 132, in build
    raise ValueError('The channel dimension of the inputs '
ValueError: The channel dimension of the inputs should be defined. Found `None`.

from ISR.models import RDN
#import keras.backend

import numpy as np
from PIL import Image

#img = Image.open('/home/steven/Downloads/super-resolution/Images/section8-image.png')
img = Image.open('/home/steven/Downloads/super-resolution/Images/Living.png')
#img = img.convert("RGB")
lr_img = np.array(img)

rdn = RDN(arch_params={'C':6, 'D':20, 'G':64, 'G0':64, 'x':2})

rdn.model.load_weights('/home/steven/Downloads/super-resolution/ISR/image-super-resolution/weights/sample_weights/rdn-C6-D20-G64-G064-x2_PSNR_epoch086.hdf5')
#rdn-C6-D20-G64-G064-x2_ArtefactCancelling_epoch219.hdf5')

sr_img = rdn.predict(lr_img)
Image.fromarray(sr_img)

image that I am trying to load:
https://paste.pics/eb292e208845d06fa588cfadc377fda5
apt install git-lfs
note git lfs pull doesn't work but I was able to download the files from
https://drive.google.com/uc?id=1u9ituA3ScttN9Vi-UkALmpO0dWQLm8Rv&export=download and
https://github.com/idealo/image-super-resolution/tree/master/weights/sample_weights


Thanks in advance for your help!

**Comment by amughrabi** [comment]
> Edit:
> I added a
> `img = img.convert("RGB")`

I have the same problem, and I solved it by converting the PIL image to cv2 image. See below:
```python
from PIL import Image
import cv2
import numpy as np


def to_cv2_image(img_pil):
    img = np.array(img_pil)
    mode = cv2.COLOR_RGB2BGR
    if has_alpha_channel(img_pil):
        mode = cv2.COLOR_RGBA2BGR
    return cv2.cvtColor(img, mode)


def has_alpha_channel(img_pil):
    """
    check if the image has alpha, but it does not have to be transparent
    :param img_pil: PIL image
    :return: True if the image has alpha channel, otherwise false.
    """
    return img_pil.mode == 'RGBA'

if __name__ == '__main__':
     img = Image.open("/path/to/image")
     img = to_cv2_image(img)
     img = Image.fromarray(img)
     img = np.array(img)
     img = self.rrdn.predict(img)
     img = np.array(img)
     img = Image.fromarray(img)
     img = to_cv2_image(img)
```
